File: frontend/src/lib/services/skill-analyzer.ts
import yaml from "js-yaml"
import { SKILL_TEMPLATES } from "@/lib/data/templates"
import type {
  CompletenessReport,
  ContentQualityReport,
  FileStructureReport,
  FrontmatterReport,
  NameValidation,
  SkillFrontmatter,
  SkillTemplate,
} from "@/types/skill"

/* ------------------------------------------------------------------ */
/*  Constants                                                          */
/* ------------------------------------------------------------------ */

const KNOWN_FIELDS = new Set(["name", "description", "license", "compatibility", "metadata"])

const VALID_NAME_RE = /^[a-z0-9]+(-[a-z0-9]+)*$/
const NAME_MIN = 1
const NAME_MAX = 64

const SECTION_TRIGGER_RE = /#+\s*(trigger|when\s+to\s+use|usage|when\s+this\s+skill)/i

/* ------------------------------------------------------------------ */
/*  Name validation                                                    */
/* ------------------------------------------------------------------ */

function validateName(name: string | null): NameValidation {
  if (!name) return { valid: false, value: null, issues: ["missing name"] }
  const issues: string[] = []
  if (name.length < NAME_MIN) issues.push(`name must be at least ${NAME_MIN} character(s)`)
  if (name.length > NAME_MAX) issues.push(`name must be at most ${NAME_MAX} characters (got ${name.length})`)
  if (name[0] === "-" || name[name.length - 1] === "-") issues.push("name must not start or end with '-'")
  if (name.includes("--")) issues.push("name must not contain consecutive '--'")
  if (!VALID_NAME_RE.test(name)) issues.push("name must be lowercase alphanumeric with single hyphen separators")
  return { valid: issues.length === 0, value: name, issues }
}

/* ------------------------------------------------------------------ */
/*  Frontmatter parsing                                                */
/* ------------------------------------------------------------------ */

function parseFrontmatter(text: string): FrontmatterReport {
  const emptyFields: SkillFrontmatter = {
    name: null,
    description: null,
    license: null,
    compatibility: null,
    metadata: null,
  }

  const stripped = text.trimStart()
  if (!stripped.startsWith("---")) {
    return { present: false, valid_yaml: false, fields: emptyFields }
  }

  const secondDelim = stripped.indexOf("---", 3)
  if (secondDelim === -1) {
    return {
      present: true,
      valid_yaml: false,
      parse_error: "Unterminated frontmatter: missing closing ---",
      fields: emptyFields,
    }
  }

  const rawYaml = stripped.slice(3, secondDelim).trim()
  if (!rawYaml) {
    return {
      present: true,
      valid_yaml: false,
      parse_error: "Empty frontmatter block",
      fields: emptyFields,
    }
  }

  let parsed: unknown
  try {
    parsed = yaml.load(rawYaml)
  } catch (e) {
    return {
      present: true,
      valid_yaml: false,
      parse_error: `YAML parse error: ${e instanceof Error ? e.message : e}`,
      fields: emptyFields,
    }
  }

  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    return {
      present: true,
      valid_yaml: false,
      parse_error: `Frontmatter must be a YAML mapping, got ${Array.isArray(parsed) ? "array" : typeof parsed}`,
      fields: emptyFields,
    }
  }

  const d = parsed as Record<string, unknown>

  let metadata: Record<string, string> | null = null
  const rawMeta = d.metadata
  if (rawMeta && typeof rawMeta === "object" && !Array.isArray(rawMeta)) {
    metadata = Object.fromEntries(Object.entries(rawMeta as Record<string, unknown>).map(([k, v]) => [k, String(v)]))
  }

  const fields: SkillFrontmatter = {
    name: "name" in d ? String(d.name) : null,
    description: "description" in d ? String(d.description) : null,
    license: "license" in d ? String(d.license) : null,
    compatibility: "compatibility" in d ? String(d.compatibility) : null,
    metadata,
  }

  const missingRequired: string[] = []
  if (fields.name === null) missingRequired.push("name")
  if (fields.description === null) missingRequired.push("description")

  const unknownFields = Object.keys(d).filter((k) => !KNOWN_FIELDS.has(k))

  const nameVal = validateName(fields.name)

  return {
    present: true,
    valid_yaml: true,
    fields,
    name_validation: nameVal,
    missing_required: missingRequired,
    unknown_fields: unknownFields,
  }
}

/* ------------------------------------------------------------------ */
/*  Content analysis                                                   */
/* ------------------------------------------------------------------ */

function analyzeContent(text: string): ContentQualityReport {
  const stripped = text.trimStart()
  let content = ""

  if (stripped.startsWith("---")) {
    const secondDelim = stripped.indexOf("---", 3)
    if (secondDelim !== -1) content = stripped.slice(secondDelim + 3).trim()
  }

  if (!content) {
    return {
      has_content: false,
      word_count: 0,
      sections: [],
      has_when_to_use: false,
      has_examples: false,
      issues: ["No content after frontmatter — skill has no instructions"],
      score: 0,
    }
  }

  const words = content.split(/\s+/)
  const wordCount = words.length

  const headingMatches = content.match(/^#{1,6}\s+(.+)$/gm)
  const headings = headingMatches ? headingMatches.map((h) => h.replace(/^#+\s+/, "")) : []

  const hasWhenToUse = SECTION_TRIGGER_RE.test(content)
  const hasExamples = content.includes("```") || content.toLowerCase().includes("example")

  const issues: string[] = []
  let score = 40

  if (wordCount < 50) {
    issues.push("Content is very short (<50 words) — consider adding more detail")
  } else if (wordCount >= 200) {
    score += 20
  } else if (wordCount >= 100) {
    score += 10
  }

  if (headings.length === 0) {
    issues.push("No headings found — structure your skill with sections")
  } else if (headings.length >= 3) {
    score += 10
  }

  if (!hasWhenToUse) {
    issues.push("No trigger/usage section found — add a section describing when the skill should be loaded")
  } else {
    score += 15
  }

  if (!hasExamples) {
    issues.push("No code examples found — examples help agents understand the skill")
  } else {
    score += 15
  }

  score = Math.max(0, Math.min(100, score))

  return {
    has_content: true,
    word_count: wordCount,
    sections: headings,
    has_when_to_use: hasWhenToUse,
    has_examples: hasExamples,
    issues,
    score,
  }
}

/* ------------------------------------------------------------------ */
/*  File structure check                                               */
/* ------------------------------------------------------------------ */

function analyzeFileStructure(filename: string): FileStructureReport {
  const issues: string[] = []
  const filenameOk = filename === "SKILL.md"
  if (!filenameOk) issues.push(`File should be named SKILL.md (got ${filename})`)
  return { filename, filename_ok: filenameOk, directory_matches_name: null, issues }
}

/* ------------------------------------------------------------------ */
/*  Full analysis                                                      */
/* ------------------------------------------------------------------ */

export function analyzeSkill(raw: string, filename = "SKILL.md"): CompletenessReport {
  const fmReport = parseFrontmatter(raw)
  const contentReport = analyzeContent(raw)
  const fileReport = analyzeFileStructure(filename)

  let score = 0
  if (fmReport.valid_yaml) {
    score += 30
    if (!fmReport.missing_required || fmReport.missing_required.length === 0) score += 15
    if (fmReport.name_validation?.valid) score += 15
  } else if (fmReport.present) {
    score += 10
  }

  score += Math.floor(contentReport.score / 2)
  score = Math.max(0, Math.min(100, score))

  const parts: string[] = []
  if (fmReport.valid_yaml && (!fmReport.missing_required || fmReport.missing_required.length === 0) && fmReport.name_validation?.valid) {
    parts.push("Frontmatter is valid and complete")
  } else if (fmReport.present) {
    parts.push("Frontmatter needs fixes")
  } else {
    parts.push("Frontmatter is missing")
  }

  if (contentReport.score >= 60) {
    parts.push("content is well-structured")
  } else if (contentReport.has_content) {
    parts.push("content needs improvement")
  } else {
    parts.push("no content")
  }

  return {
    overall_score: score,
    summary: parts.join("; "),
    frontmatter: fmReport,
    content_quality: contentReport,
    file_structure: fileReport,
  }
}

/* ------------------------------------------------------------------ */
/*  Templates                                                          */
/* ------------------------------------------------------------------ */

export function getTemplates(): SkillTemplate[] {
  return SKILL_TEMPLATES
}

File: backend/app/services/skill_analyzer.py
from __future__ import annotations

import re
from typing import Any

import yaml

from app.data.templates import SKILL_TEMPLATES
from app.schemas.skill import (
    CompletenessReport,
    ContentQualityReport,
    FileStructureReport,
    FrontmatterReport,
    NameValidation,
    SkillFrontmatter,
    SkillTemplate,
)

KNOWN_FRONTMATTER_FIELDS = {"name", "description", "license", "compatibility", "metadata"}

VALID_NAME_RE = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*$")
NAME_MIN_LEN = 1
NAME_MAX_LEN = 64

SECTION_TRIGGER_PATTERNS = re.compile(
    r"#+\s*(trigger|when\s+to\s+use|usage|when\s+this\s+skill)",
    re.IGNORECASE,
)


def _validate_name(name: str | None) -> NameValidation:
    if not name:
        return NameValidation(valid=False, value=None, issues=["missing name"])
    issues: list[str] = []
    if len(name) < NAME_MIN_LEN:
        issues.append(f"name must be at least {NAME_MIN_LEN} character(s)")
    if len(name) > NAME_MAX_LEN:
        issues.append(f"name must be at most {NAME_MAX_LEN} characters (got {len(name)})")
    if name[0] == "-" or name[-1] == "-":
        issues.append("name must not start or end with '-'")
    if "--" in name:
        issues.append("name must not contain consecutive '--'")
    if not VALID_NAME_RE.match(name):
        issues.append("name must be lowercase alphanumeric with single hyphen separators")
    return NameValidation(valid=len(issues) == 0, value=name, issues=issues)


def _parse_frontmatter(text: str) -> FrontmatterReport:
    """Extract and validate YAML frontmatter from SKILL.md text."""
    stripped = text.lstrip()
    if not stripped.startswith("---"):
        return FrontmatterReport(present=False, valid_yaml=False, fields=SkillFrontmatter())

    second = stripped.find("---", 3)
    if second == -1:
        return FrontmatterReport(
            present=True,
            valid_yaml=False,
            parse_error="Unterminated frontmatter: missing closing ---",
            fields=SkillFrontmatter(),
        )

    raw_yaml = stripped[3:second].strip()
    if not raw_yaml:
        return FrontmatterReport(
            present=True,
            valid_yaml=False,
            parse_error="Empty frontmatter block",
            fields=SkillFrontmatter(),
        )

    try:
        parsed: Any = yaml.safe_load(raw_yaml)
    except yaml.YAMLError as e:
        return FrontmatterReport(
            present=True,
            valid_yaml=False,
            parse_error=f"YAML parse error: {e}",
            fields=SkillFrontmatter(),
        )

    if not isinstance(parsed, dict):
        return FrontmatterReport(
            present=True,
            valid_yaml=False,
            parse_error=f"Frontmatter must be a YAML mapping, got {type(parsed).__name__}",
            fields=SkillFrontmatter(),
        )

    metadata: dict[str, str] | None = None
    raw_meta = parsed.get("metadata")
    if isinstance(raw_meta, dict):
        metadata = {str(k): str(v) for k, v in raw_meta.items()}

    fields = SkillFrontmatter(
        name=str(parsed["name"]) if "name" in parsed else None,
        description=str(parsed["description"]) if "description" in parsed else None,
        license=str(parsed.get("license")) if "license" in parsed else None,
        compatibility=str(parsed.get("compatibility")) if "compatibility" in parsed else None,
        metadata=metadata,
    )

    missing_required = []
    if fields.name is None:
        missing_required.append("name")
    if fields.description is None:
        missing_required.append("description")

    unknown_fields = [k for k in parsed if k not in KNOWN_FRONTMATTER_FIELDS]

    name_val = _validate_name(fields.name)

    return FrontmatterReport(
        present=True,
        valid_yaml=True,
        fields=fields,
        name_validation=name_val,
        missing_required=missing_required,
        unknown_fields=unknown_fields,
    )


def _analyze_content(text: str) -> ContentQualityReport:
    """Analyze the markdown content after frontmatter."""
    stripped = text.lstrip()
    content = ""
    if stripped.startswith("---"):
        second = stripped.find("---", 3)
        if second != -1:
            content = stripped[second + 3 :].strip()

    if not content:
        return ContentQualityReport(
            has_content=False,
            word_count=0,
            sections=[],
            has_when_to_use=False,
            has_examples=False,
            issues=["No content after frontmatter — skill has no instructions"],
            score=0,
        )

    words = content.split()
    word_count = len(words)

    sections = re.findall(r"^(#{1,6})\s+(.+)$", content, re.MULTILINE)
    headings = [h[1] for h in sections]

    has_when_to_use = bool(SECTION_TRIGGER_PATTERNS.search(content))
    has_examples = "```" in content or "example" in content.lower()

    issues: list[str] = []
    score = 40

    if word_count < 50:
        issues.append("Content is very short (<50 words) — consider adding more detail")
    elif word_count >= 200:
        score += 20
    elif word_count >= 100:
        score += 10

    if len(headings) == 0:
        issues.append("No headings found — structure your skill with sections")
    elif len(headings) >= 3:
        score += 10

    if not has_when_to_use:
        issues.append(
            "No trigger/usage section found — "
            "add a section describing when the skill should be loaded"
        )
    else:
        score += 15

    if not has_examples:
        issues.append("No code examples found — examples help agents understand the skill")
    else:
        score += 15

    score = max(0, min(100, score))

    return ContentQualityReport(
        has_content=True,
        word_count=word_count,
        sections=headings,
        has_when_to_use=has_when_to_use,
        has_examples=has_examples,
        issues=issues,
        score=score,
    )


def _analyze_file_structure(filename: str, expected_dir_name: str | None) -> FileStructureReport:
    """Check file structure conventions."""
    issues: list[str] = []
    filename_ok = filename == "SKILL.md"
    if not filename_ok:
        issues.append(f"File should be named SKILL.md (got {filename})")

    dir_matches = None
    if expected_dir_name:
        dir_matches = filename_ok and expected_dir_name == expected_dir_name  # pass-through
    # Note: we can't verify directory name from just the file content
    # The user would need to provide it separately

    return FileStructureReport(
        filename=filename,
        filename_ok=filename_ok,
        directory_matches_name=dir_matches,
        issues=issues,
    )


def analyze_skill(raw: str, filename: str = "SKILL.md") -> CompletenessReport:
    """Full SKILL.md analysis."""
    fm_report = _parse_frontmatter(raw)
    content_report = _analyze_content(raw)
    file_report = _analyze_file_structure(filename, fm_report.fields.name)

    score = 0
    if fm_report.valid_yaml:
        score += 30
        if not fm_report.missing_required:
            score += 15
        if fm_report.name_validation.valid:
            score += 15
    elif fm_report.present:
        score += 10

    score += content_report.score // 2
    score = max(0, min(100, score))

    parts: list[str] = []
    if fm_report.valid_yaml and not fm_report.missing_required and fm_report.name_validation.valid:
        parts.append("Frontmatter is valid and complete")
    elif fm_report.present:
        parts.append("Frontmatter needs fixes")
    else:
        parts.append("Frontmatter is missing")

    if content_report.score >= 60:
        parts.append("content is well-structured")
    elif content_report.has_content:
        parts.append("content needs improvement")
    else:
        parts.append("no content")

    summary = "; ".join(parts)

    return CompletenessReport(
        overall_score=score,
        summary=summary,
        frontmatter=fm_report,
        content_quality=content_report,
        file_structure=file_report,
    )


def get_templates() -> list[SkillTemplate]:
    """Return the list of built-in skill templates."""
    return SKILL_TEMPLATES
